pkgs/dart_mcp/lib/src/api/tools.dart

```diff
@@ -329,34 +709,297 @@ extension type ObjectSchema.fromMap(Map<String, Object?> _value)
+  /// a lowercase letter:
+  ///
+  /// ```dart
+  /// final schema = ObjectSchema(
+  ///   propertyNames: Schema.string(pattern: r'^[a-z].*$'),
+  /// );
+  /// ```
+  ///
+  /// In this schema, an object like `{'name': 'John', 'age': 30}` would be
+  /// valid, but `{'Name': 'John', 'Age': 30}` would be invalid because the
+  /// property names do not start with a lowercase letter.
   StringSchema? get propertyNames =>
       (_value['propertyNames'] as Map?)?.cast<String, Object?>()
           as StringSchema?;
 
   /// The minimum number of properties in this object.
+  ///
+  /// If the object has less than this many properties, it will be invalid.
   int? get minProperties => _value['minProperties'] as int?;
 
   /// The maximum number of properties in this object.
+  ///
+  /// If the object has more than this many properties, it will be invalid.
   int? get maxProperties => _value['maxProperties'] as int?;
+
+  bool _validateObject(
+    Object? data,
+    List<String> currentPath,
+    List<ValidationError> accumulatedFailures,
+  ) {
+    if (data is! Map<String, Object?>) {
+      accumulatedFailures.add(
+        ValidationError(ValidationErrorType.typeMismatch, path: currentPath),
+      );
+      return false;
+    }
+
+    var isValid = true;
+
+    if (minProperties case final min? when data.keys.length < min) {
+      isValid = false;
+      accumulatedFailures.add(
+        ValidationError(
+          ValidationErrorType.minPropertiesNotMet,
+          path: currentPath,
+          details:
+              'There should be at least $minProperties '
+              'properties. Only ${data.keys.length} were found.',
+        ),
+      );
+    }
+
+    if (maxProperties case final max? when data.keys.length > max) {
+      isValid = false;
+      accumulatedFailures.add(
+        ValidationError(
+          ValidationErrorType.maxPropertiesExceeded,
+          path: currentPath,
+          details:
+              'Exceeded maxProperties limit of $maxProperties '
+              '(${data.keys.length}).',
+        ),
+      );
+    }
+
+    for (final reqProp in required ?? const []) {
+      if (!data.containsKey(reqProp)) {
+        isValid = false;
+        accumulatedFailures.add(
+          ValidationError(
+            ValidationErrorType.requiredPropertyMissing,
+            path: currentPath,
+            details: 'Required property "$reqProp" is missing.',
+          ),
+        );
+      }
+    }
+
+    // Check for property values that match any defined properties in the
+    // `properties` map. If a property name exists in the `properties` map, the
+    // value of that property is validated against the schema associated with
+    // that property.
+    final evaluatedKeys = <String>{};
+    if (properties case final props?) {
+      for (final entry in props.entries) {
+        if (data.containsKey(entry.key)) {
+          evaluatedKeys.add(entry.key);
+          final propertyPath = [...currentPath, entry.key];
```

Comment:
We may want to consider just adding to the currentPath and then removing from it instead of copying all the way down. And then just do the copy when creating an error object.

Reply:
Done.  I now only copy it when I create a new `ValidationError`.